solver/solver.go

```diff
@@ -26,30 +26,35 @@ import (
 )
 
 type LLBOpt struct {
-	SourceManager    *source.Manager
-	CacheManager     cache.Manager
-	Worker           worker.Worker
-	InstructionCache InstructionCache
-	ImageSource      source.Source
+	WorkerController *worker.Controller
 	Frontends        map[string]frontend.Frontend // used by nested invocations
-	CacheExporter    *cacheimport.CacheExporter
-	CacheImporter    *cacheimport.CacheImporter
+}
+
+// DetermineVertexWorker determines worker for a vertex.
+// Currently, constraint is just ignored.
+func DetermineVertexWorker(wc *worker.Controller, v Vertex) (*worker.Worker, error) {
+	// TODO: multiworker
+	return wc.GetDefault()
 }
 
 func NewLLBSolver(opt LLBOpt) *Solver {
 	var s *Solver
 	s = New(func(v Vertex) (Op, error) {
+		w, err := DetermineVertexWorker(opt.WorkerController, v)
```

Comment:
follow-up: worker can't be determined here, both constraints and the workers of the inputs are needed.